aptos-move/aptos-vm-types/src/change_set.rs

```diff
@@ -60,71 +175,167 @@ impl VMChangeSet {
+        // storage.
+        let mut resource_write_set = StateChange::empty();
+        let mut module_write_set = StateChange::empty();
+
+        // Go over write ops to split them into resources and modules.
+        for (state_key, write_op) in write_set {
+            if matches!(state_key.inner(), StateKeyInner::AccessPath(ap) if ap.is_code()) {
+                module_write_set.insert(state_key, write_op);
+            } else {
+                // Everything else must be a resource.
+                // TODO(aggregator): Aggregator V1 is a table item, and so fits into
+                // this category. In practice this should never happen, but we might
+                // want to have an assert here before Aggregator V2 lands.
+                resource_write_set.insert(state_key, write_op);
+            }
+        }
+
+        let change_set = Self {
+            resource_write_set,
+            module_write_set,
+            aggregator_write_set: StateChange::empty(),
+            delta_change_set: DeltaChangeSet::empty(),
+            events,
+        };
+        checker.check_change_set(&change_set)?;
+        Ok(change_set)
+    }
+
+    /// Converts VM-native change set into its storage representation with fully
+    /// serialized changes.
+    /// At this points, if deltas are not materialized the conversion fails.
+    pub fn try_into_storage_change_set(self) -> anyhow::Result<StorageChangeSet, VMStatus> {
+        let Self {
+            resource_write_set,
+            module_write_set,
+            aggregator_write_set,
+            delta_change_set,
+            events,
+        } = self;
+
+        // Only change set without deltas can be converted.
+        if !delta_change_set.is_empty() {
+            return Err(VMStatus::error(
+                StatusCode::DATA_FORMAT_ERROR,
+                err_msg(
+                    "Cannot convert from VMChangeSet with non-materialized deltas to ChangeSet.",
+                ),
+            ));
+        }
+
+        // Create a write set which has all the changes combined together.
+        let mut write_set_mut = WriteSetMut::default();
+        write_set_mut.extend(resource_write_set);
+        write_set_mut.extend(module_write_set);
+
+        // TODO(aggregator): Aggregator V1 is a state item, and has to be added
+        // here.
+        // When we move to different key representation for aggregators,
+        // this has to be conditional on the table item (supply) until we
+        // fully transition to the new version.
+        write_set_mut.extend(aggregator_write_set);
+
+        let write_set = write_set_mut.freeze().map_err(|_| {
+            VMStatus::error(
+                StatusCode::DATA_FORMAT_ERROR,
+                err_msg("Failed to freeze write sets when converting VMChangeSet to ChangeSet"),
+            )
+        })?;
+        Ok(StorageChangeSet::new(write_set, events))
+    }
+
+    /// A useful method to iterate over all writes.
```

Comment:
nit: remove :)
all methods should be `Useful`, and the name clearly conveys what the method does.